### PR TITLE
User Defined Plant Component Refactor

### PR DESCRIPTION
**Pull request by mitchute**
### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mitchute** [comment]
Ready for review

**Comment by Myoldmopar** [comment]
I merged develop in and addressed the tiny conflict along the way, I'll review the code now and let CI work on it overnight to make sure my conflict resolution wasn't correct.

File: src/EnergyPlus/Plant/PlantManager.cc [hunk 1]
```diff
@@ -1440,6 +1441,7 @@ namespace EnergyPlus {
                                 this_comp.TypeOf_Num = TypeOf_PlantComponentUserDefined;
                                 this_comp.GeneralEquipType = GenEquipTypes_PlantComponent;
                                 this_comp.CurOpSchemeType = UnknownStatusOpSchemeType;
+                                this_comp.compPtr = UserDefinedComponents::UserPlantComponentStruct::factory(CompNames(CompNum));
```
Comment (Myoldmopar) [inline src/EnergyPlus/Plant/PlantManager.cc:1444]: Simple enough.

File: src/EnergyPlus/PlantLoopEquip.cc [hunk 2]
```diff
@@ -934,15 +933,7 @@ namespace PlantLoopEquip {
...
-
-                SimUserDefinedPlantComponent(
-                    LoopNum, LoopSideNum, sim_component.TypeOf, sim_component.Name, EquipNum, InitLoopEquip, CurLoad, MaxLoad, MinLoad, OptLoad);
-                if (InitLoopEquip) {
-                    sim_component.MaxLoad = MaxLoad;
-                    sim_component.MinLoad = MinLoad;
-                    sim_component.OptLoad = OptLoad;
-                    sim_component.CompNum = EquipNum;
-                }
+                sim_component.compPtr->simulate(sim_component_location, FirstHVACIteration, CurLoad, RunFlag);
```
Comment (Myoldmopar) [inline src/EnergyPlus/PlantLoopEquip.cc:936]: All good here.

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc [hunk 3]
```diff
@@ -464,6 +466,7 @@ void EnergyPlusFixture::clear_all_states()
     UnitarySystems::clear_state();
     UnitHeater::clear_state();
     UnitVentilator::clear_state();
+    UserDefinedComponents::clear_state();
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:469]: Excellent!

File: src/EnergyPlus/UserDefinedComponents.hh [hunk 4]
```diff
@@ -199,11 +183,24 @@ namespace UserDefinedComponents {
...
+        void onInitLoopEquip(const PlantLocation &calledFromLocation) override;
+
+        void getDesignCapacities(const PlantLocation &calledFromLocation, Real64 &MaxLoad, Real64 &MinLoad, Real64 &OptLoad) override;
+
+        void simulate(const PlantLocation &calledFromLocation, bool FirstHVACIteration, Real64 &CurLoad, bool RunFlag) override;
+
+        void initialize(int LoopNum, Real64 MyLoad);
+
+        void report(int LoopNum);
     };
```
Comment (Myoldmopar) [inline src/EnergyPlus/UserDefinedComponents.hh:204]: Nicely done with clean function names. :+1:

File: src/EnergyPlus/UserDefinedComponents.cc [hunk 5]
```diff
@@ -128,30 +101,109 @@ namespace UserDefinedComponents {
...
+        UserCoil.deallocate();
+        UserZoneAirHVAC.deallocate();
+        UserAirTerminal.deallocate();
+    }
+
+    PlantComponent *UserPlantComponentStruct::factory(std::string const &objectName)
+    {
+        // Process the input data
+        if (GetPlantCompInput) {
+            GetUserDefinedPlantComponents();
```
Comment (Myoldmopar) [inline src/EnergyPlus/UserDefinedComponents.cc:135]: Right, just getting plant objects, great.

File: src/EnergyPlus/UserDefinedComponents.cc [hunk 6]
```diff
@@ -128,30 +101,109 @@ namespace UserDefinedComponents {
...
+        Real64 myLoad = 0.0;
+        int thisLoop = 0;
+
+        this->initialize(calledFromLocation.loopNum, myLoad);
+
+        for (int loop = 1; loop <= this->NumPlantConnections; ++loop) {
+            if (calledFromLocation.loopNum != this->Loop(loop).LoopNum) continue;
+            if (calledFromLocation.loopSideNum != this->Loop(loop).LoopSideNum) continue;
+            thisLoop = loop;
+        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/UserDefinedComponents.cc:162]: This "feels" like an odd way to do this, but I'm sure it's just pulled from the original code, so no need to change anything, just unusual. :man_shrugging:

File: src/EnergyPlus/UserDefinedComponents.cc [hunk 7]
```diff
@@ -2336,121 +2175,86 @@ namespace UserDefinedComponents {
...
+                                                                              DataLoopNode::Node(this->Loop(loop).InletNodeNum).Temp,
+                                                                              DataPlant::PlantLoop(this->Loop(loop).LoopNum).FluidIndex,
+                                                                              RoutineName);
+                this->Loop(loop).InletCp = FluidProperties::GetSpecificHeatGlycol(DataPlant::PlantLoop(this->Loop(loop).LoopNum).FluidName,
+                                                                                  DataLoopNode::Node(this->Loop(loop).InletNodeNum).Temp,
+                                                                                  DataPlant::PlantLoop(this->Loop(loop).LoopNum).FluidIndex,
+                                                                                  RoutineName);
+                this->Loop(loop).InletTemp = DataLoopNode::Node(this->Loop(loop).InletNodeNum).Temp;
+                this->Loop(loop).InletMassFlowRate = DataLoopNode::Node(this->Loop(loop).InletNodeNum).MassFlowRate;
```
Comment (Myoldmopar) [inline src/EnergyPlus/UserDefinedComponents.cc:2252]: This is all a nice cleanup.

**Review by Myoldmopar — APPROVED** [review]
Everything looks good on the code side.

**Comment by Myoldmopar** [comment]
If CI is happy, this can merge in the morning.  Thanks, @mitchute

**Pull request merged**